source/common/router/retry_state_impl.cc

```diff
@@ -186,6 +189,11 @@ RetryStatus RetryStateImpl::shouldRetry(const Http::HeaderMap* response_headers,
 }
 
 bool RetryStateImpl::wouldRetryFromHeaders(const Http::HeaderMap& response_headers) {
+  // We should retry rate limited request only if the policy is set.
+  if (Http::CodeUtility::isRateLimited(Http::Utility::getResponseStatus(response_headers))) {
```

Comment:
@mattklein123 This is the check I am talking about. It checks if the request is rate limited and the retry policy for Retry_on_Rate_Limited is set then only retries. Otherwise just does not retry and check for other gRPC conditions. Am I understanding this incorrectly?

Reply:
If you think opt-in behaviour is not needed, we could some thing similar to https://github.com/envoyproxy/envoy/blob/master/source/common/router/retry_state_impl.cc#L266??

Reply:
How does this help with gRPC though? In a normal gRPC response the status code is going to be 200, so this wouldn't apply?

Reply:
I was not intending this for normal gRPC. I was intending this for a use case where the rate limiting filter identifies the request as rate limited and that returns status code of 429. Do you think this configuration is generic and people get confused? That is why in the initial PR I added the clause that `returned by rate limiting service` in the doc explanation. Do you see any other way of achieving the same?

Reply:
Sorry, stepping back, can you explain what you are trying to achieve? You originally said you don't want to retry rate limited requests, but this PR will allow you to do that? I don't see any functional change in this PR other than an addition of a new rate limit policy which I think you said you don't want to use?

Reply:
Ok. Here is what we want to do. We enable rate limiting filter. So it returns 429 status code and `UNAVAILABLE` gRPC code. We have set retry-grpc-on as `UNAVAILABLE` in addition to other conditions. So because of this the rate limited requests are retried which we do not want.
So in this PR, I am ensuring that if a request is rate limited (status 429), it will not retry unless you provide the new policy (rate-limited). So for us, assuming that  this PR is merged with out adding any new retry policies it would stop retrying because of the added condition. The new policy would just be opt-in for who ever wants to retry rate limited requests.

Thinking further, does this policy have to handle normal gRPC? Should n't the normal gRPC go through retry-on-grpc route?

Reply:
Is your caller gRPC or HTTP?

There is this PR: https://github.com/envoyproxy/envoy/pull/4879. This will allow you to change Envoy's gRPC mapping of 429 to resource exhausted. If your caller is gRPC, this might be the only thing you need?

I think the confusion here is that for a gRPC caller, HTTP status is converted to 200 (OK), so what you are doing here is not going to help for gRPC. If you have an HTTP caller, 429 is not retried by default, so you wouldn't get a retry either.

Basically, I don't think we need this PR for your use case?

Reply:
@mattklein123  Our caller is gRPC. I am following that PR of mapping change. However, we see some cases where RESOURCES_EXHAUSTED is returned by the service (other than rate limiting). So ideally we would want the RESOURCE_EXHAUSTED to in the retry policy but exclude rate limiting calls.

>I think the confusion here is that for a gRPC caller, HTTP status is converted to 200 (OK), so what you are doing here is not going to help for gRPC. If you have an HTTP caller, 429 is not retried by default, so you wouldn't get a retry either.
Thanks for this. This is creating the confusion for me. I did not know that for gRPC HTTP status is converted to 200(OK). I was missing that point.  Agree this PR is not going to help for that case. Sorry for all the confusion. 

So basically if we want to have `RESOURCE_EXHAUSTED` in the retry codes list (after #4879 is merged) and do not want to retry rate limited requests - how should we proceed? 
Should we add a new policy as proposed here (https://github.com/envoyproxy/envoy/issues/4855#event-1935500673 - option 2)? or any other better solutions?

Reply:
I guess if no code changes are possible, that would be best. For example, do you really need to retry in the case of RESOURCE_EXHAUSTED? That seems generally bad.

If you do, I guess setting a header much like we do for overloaded makes sense to me.

Reply:
@mattklein123 just to confirm, if we implement this new header approach, do you think it is better to configure the behaviour of enforcing retries based on some policy or just do not retry (and document this behaviour) if we see that header exactly as how `x-envoy-overloaded` works?

Reply:
I would rather not add a retry policy until someone explicitly asks for it?

Reply:
Makes sense..I agree with it. I have pushed the PR https://github.com/envoyproxy/envoy/pull/4972 with above changes. PTAL.